File: trials/linked_list.c
//Author: Sanskriti Gupta
//Date: 10 March 2022
//Aim: To implement insertion/deletion at beginning, end and position of a singly linked list

//YE WALA SAHI HAI !!!!!

#include<stdlib.h>
#include <stdio.h>
struct Node 
{
	int data;
	struct Node *link;
};
struct Node *start=NULL;
void display()
{
    struct Node *ptr=start;
    while(ptr!=NULL)
    {
        printf("%d ",ptr->data);
        ptr=ptr->link;
    }
}
struct Node *create_node()
{
    struct Node *temp=(struct Node *)malloc(sizeof(struct Node));
    return temp;
}
void ins_beg(int item)
{
    struct Node *new=create_node();
    new->data=item;
    if(start==NULL)
    {
        start=new;
        new->link=NULL;
    }
    else
    {
        new->link=start;
        start=new;
    }
}
void ins_end(int item)
{
    struct Node *new=create_node();
    new->data=item;
    new->link=NULL;
    if(start==NULL)
    {
        start=new;
    }
    else
    {
        struct Node *ptr=start;
        while(ptr->link!=NULL)
        {
            ptr=ptr->link;
        }
        ptr->link=new;
    }
}
void ins_pos(int item, int pos)
{
    int p=1;
    struct Node *new=create_node();
    new->data=item;
    if(start==NULL)
    {
        start=new;
        new->link=NULL;
    }
    else
    {
        struct Node *ptr=start;
        while(p<pos-1)
        {
            ptr=ptr->link;
            p++;
        }
        new->link=ptr->link;
        ptr->link=new;
    }
}
int del_beg()
{
    if(start==NULL)
    {
        printf("\nEmpty List");
    }
    else
    {
        int i=start->data;
        start=start->link;
        return i;
    }
}
int del_end()
{
    if(start==NULL)
    {
        printf("\nList is Empty");
    }
    else
    {
        struct Node *ptr=start;
        while(ptr->link->link!=NULL)
        {
            ptr=ptr->link;
        }
        int i=ptr->link->data;
        struct Node *temp=ptr->link;
        ptr->link=NULL;
        free(temp);
        return i;
    }
}
int del_pos(int pos)
{
    if(start==NULL)
    {
        printf("\nList is Empty");
    }
    else
    {
        int p=1;
        struct Node *ptr=start;
        while(p<pos-1)
        {
            ptr=ptr->link;
            p++;
        }
        int i=ptr->link->data;
        struct Node *temp=ptr->link;
        ptr->link=ptr->link->link;
        free(temp);
        return i;
    }
}
void main()
{
    int item,pos,choice;
    while(1)
    {
        printf("\n1. Insert at bginning");
        printf("\n2. Insert at end");
        printf("\n3. Insert at position");
        printf("\n4. Delete at beginning");
        printf("\n5. Delete at end");
        printf("\n6. Delete at position");
        printf("\n7. Display");
        printf("\n8. Exit\n");
        scanf("%d",&choice);
        switch(choice)
        {
            case 1:
            printf("\nEnter the element: ");
            scanf("%d",&item);
            ins_beg(item);
            break;
            case 2:
            printf("\nEnter the element: ");
            scanf("%d",&item);
            ins_end(item);
            break;
            case 3:
            printf("\nEnter the element: ");
            scanf("%d",&item);
            printf("\nEnter the position: ");
            scanf("%d",&pos);
            ins_pos(item,pos);
            break;
            case 4:
            item=del_beg();
            printf("\nDeleted Item: %d",item);
            break;
            case 5:
            item=del_end();
            printf("\nDeleted Item: %d",item);
            break;
            case 6:
            printf("\nEnter the position: ");
            scanf("%d",&pos);
            item=del_pos(pos);
            printf("\nDeleted Item: %d",item);
            break;
            case 7:
            display();
            break;
            case 8:
            exit(0);
            default:
            printf("\nWrong Choice");
        }
    }
}

File: practice/linkedlist.java
//package practice;
import java.util.Scanner;
import java.lang.System;
public class linkedlist
{
    public static class Node
    {
        int data;
        Node next;
        Node()
        {
            this.next = null;
        }
    }
    public static Node start;
    public static Node temp;
    static class ll_functions
    {
        void ins_beg(int data)
        {
            Node n = new Node();
            n.data = data;
            if(start==null)
            {
                start = n;
            }
            else
            {
                n.next = start;
                start = n;
            }
        }
        void ins_end(int data)
        {
            Node n = new Node();
            n.data = data;
            if(start==null)
            {
                start = n;
            }
            else
            {
                temp = start;
                while(temp.next!=null)
                {
                    temp = temp.next;
                }
                temp.next = n;
            }
        }
        void ins_pos(int data, int pos)
        {
            Node n  = new Node();
            n.data = data;
            if(start==null)
            {
                start = n;
            }
            else
            {
                int pos1=0;
                temp = start;
                while(pos1 != pos-2)
                {
                    temp = temp.next;
                    pos1++;
                }
                n.next = temp.next;
                temp.next=n;
            }
        }
        int del_beg()
        {
            if(start==null)
            {
                return 0;
            }
            else
            {
                int i = start.data;
                start = start.next;
                return i;
            }
        }
        int del_end()
        {
            if(start==null)
            {
                return 0;
            }
            else
            {
                temp = start;
                while(temp.next.next!=null)
                {
                    temp = temp.next;
                }
                int i = temp.next.data;
                temp.next=null;
                return i;
            }
        }
        int del_pos(int pos)
        {
            if(start==null)
            {
                return 0;
            }
            else
            {
                temp = start;
                int pos1 = 0;
                while(pos1!=pos-2)
                {
                    temp = temp.next;
                    pos1++;
                }
                int i = temp.next.data;
                temp.next = temp.next.next;
                return i;
            }
        }
        void display()
        {
            if(start==null)
            {
                System.out.println("Empty list !!!");
            }
            else
            {
                temp = start;
                while(temp!=null)
                {
                    System.out.print(temp.data + " ");
                    temp = temp.next;
                }
            }
        }
    }
    public static void main(String[] args)
    {
        int choice, data, pos;
        Scanner sc = new Scanner(System.in);
        ll_functions ll = new ll_functions();
        while(true)
        {
            System.out.println("\n1.Insert at beg");
            System.out.println("2.Insert at end");
            System.out.println("3.Insert at position");
            System.out.println("4.Delete at beginning");
            System.out.println("5.Delete at end");
            System.out.println("6.Delete at position");
            System.out.println("7.Display");
            System.out.println("8.Exit");
            choice = sc.nextInt();
            switch(choice)
            {
                case 1:
                {
                    System.out.print("Enter the element to be inserted: ");
                    data = sc.nextInt();
                    ll.ins_beg(data);
                    break;
                }
                case 2:
                {
                    System.out.print("Enter the element to be inserted: ");
                    data = sc.nextInt();
                    ll.ins_end(data);
                    break;
                }
                case 3:
                {
                    System.out.print("Enter the element to be inserted: ");
                    data = sc.nextInt();
                    System.out.println("Enter the position: ");
                    pos = sc.nextInt();
                    ll.ins_pos(data,pos);
                    break;
                }
                case 4:
                {
                    data = ll.del_beg();
                    System.out.println("The Deleted Item: " + data);
                    break;
                }
                case 5:
                {
                    data = ll.del_end();
                    System.out.println("The Deleted Item: " + data);
                    break;
                }
                case 6:
                {
                    System.out.print("Enter the position to be deleted: ");
                    pos = sc.nextInt();
                    data = ll.del_pos(pos);
                    System.out.println("The Deleted Item: " + data);
                    break;
                }
                case 7:
                {
                    ll.display();
                    break;
                }
                case 8:
                {
                    System.exit(0);
                }
            }
        }
    }
}